src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+            }
+
+            string response = ComputeHash(ComputeHash(a1, algorithm) + Colon +
+                                        digestResponse.Parameters[Nonce] + Colon +
+                                        digestResponse.NonceCount.ToString("x8") + Colon +
+                                        cnonce + Colon +
+                                        qop + Colon +
+                                        ComputeHash(a2, algorithm), algorithm);
+
+            // Add response
+            sb.AppendKeyValue(Response, response);
+
+            // Add algorithm
+            sb.AppendKeyValue(Algorithm, algorithm, includeQuotes: false);
+
+            // Add opaque
+            sb.AppendKeyValue(Opaque, digestResponse.Parameters[Opaque]);
+
+            // Add qop
+            sb.AppendKeyValue(Qop, qop, includeQuotes: false);
+
+            // Add nc
+            sb.AppendKeyValue(NC, digestResponse.NonceCount.ToString("x8"), includeQuotes: false);
+
+            // Add cnonce
+            sb.AppendKeyValue(CNonce, cnonce, includeComma: false);
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        private static string GetRandomAlphaNumericString()
+        {
+            using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
+            {
+                const int length = 16;
+                StringBuilder sb = StringBuilderCache.Acquire(length);
+                for (int i = 0; i < length; i++)
+                {
+                    byte[] randomNumber = new byte[1];
+                    rng.GetBytes(randomNumber);
+
+                    int rangeIndex = (randomNumber[0] % 3);
+                    rng.GetBytes(randomNumber);
+
+                    if (rangeIndex == 0)
+                    {
+                        // Get a random digit 0-9
+                        sb.Append((char)(AlphaNumChooser[rangeIndex] + randomNumber[0] % 10));
+                    }
+                    else
+                    {
+                        // Get a random alphabet in a-z, A-Z
+                        sb.Append((char)(AlphaNumChooser[rangeIndex] + randomNumber[0] % 26));
+                    }
+                }
+
+                return StringBuilderCache.GetStringAndRelease(sb);
+            }
+        }
+
+        private static string ComputeHash(string data, string algorithm)
+        {
+            if (algorithm == Sha256)
+            {
+                using (HashAlgorithm hash = SHA256.Create())
+                {
+                    return ComputeHash(data, hash);
+                }
+            }
+            else
+            {
+// Disable MD5 insecure warning.
+#pragma warning disable CA5351
+                using (HashAlgorithm hash = MD5.Create())
+                {
+                    return ComputeHash(data, hash);
+                }
+#pragma warning restore CA5351
+            }
```

Comment:
Seems like this could be consolidated into something like:
```C#
#pragma warning disable CA5351 // MD5 required for protocol
using (HashAlgorithm hash = algorithm == Sha256 ? SHA256.Create() : MD5.Create())
#pragma warning restore CA5351
{
    return ComputeHash(data, hash);
}
```